bundles/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveCRC16EncapsulationCommandClass.java

```diff
@@ -0,0 +1,172 @@
+				.getNode().getNodeId());
+		int command = serialMessage.getMessagePayloadByte(offset);
+		switch (command) {
+		case 0x01: // crc 16 Encapsulation
+			handleCRC16EncapResponse(serialMessage, offset + 1);
+			break;
+		}
+	}
+
+	/**
+	 * Handle the crc16 encapsulated message. This processes the received frame,
+	 * checks the crc and forwards to the real command class.
+	 * 
+	 * @param serialMessage
+	 *            The received message
+	 * @param offset
+	 *            The starting offset into the payload
+	 */
+	private void handleCRC16EncapResponse(SerialMessage serialMessage,
+			int offset) {
+		logger.trace("Process CRC16 Encapsulation");
+
+		// Check CRC
+		byte[] payload = serialMessage.getMessagePayload();
+		byte[] messageCrc = Arrays.copyOfRange(payload, payload.length - 2,
+				payload.length);
+		byte[] tocheck = Arrays.copyOfRange(payload, offset - 2,
+				payload.length - 2);
+
+		short calculatedCrc = crc_ccit(tocheck);
+		// check if messageCrc = calculatedCrc
+		ByteBuffer byteBuffer = ByteBuffer.allocate(2);
+		byteBuffer.putShort(calculatedCrc);
+		if (!Arrays.equals(messageCrc, byteBuffer.array())) {
+			logger.error("CRC not equal");
+			return;
+		}
+
+		// Execute underlying command
+		CommandClass commandClass;
+		ZWaveCommandClass zwaveCommandClass;
+		int commandClassCode = serialMessage.getMessagePayloadByte(offset);
+		commandClass = CommandClass.getCommandClass(commandClassCode);
+		if (commandClass == null) {
+			logger.error(String.format(
+					"NODE %d: Unsupported command class 0x%02x", this.getNode()
+							.getNodeId(), commandClassCode));
+		} else {
+			zwaveCommandClass = this.getNode().getCommandClass(commandClass);
+
+			// Apparently, this node supports a command class that we did not
+			// get (yet) during initialization.
+			// Let's add it now then to support handling this message.
+			if (zwaveCommandClass == null) {
+				logger.debug(String
+						.format("NODE %d: Command class %s (0x%02x) not found, trying to add it.",
+								getNode().getNodeId(), commandClass.getLabel(),
+								commandClass.getKey()));
+
+				zwaveCommandClass = ZWaveCommandClass.getInstance(
+						commandClass.getKey(), getNode(), getController());
+
+				if (zwaveCommandClass != null) {
+					logger.debug(String.format(
+							"NODE %d: Adding command class %s (0x%02x)",
+							getNode().getNodeId(), commandClass.getLabel(),
+							commandClass.getKey()));
+					getNode().addCommandClass(zwaveCommandClass);
+				}
+			}
+
+			if (zwaveCommandClass == null) {
+				logger.error(String.format(
+						"NODE %d: CommandClass %s (0x%02x) not implemented.",
+						this.getNode().getNodeId(), commandClass.getLabel(),
+						commandClassCode));
+			} else {
+				logger.debug(String.format(
+						"NODE %d: Calling handleApplicationCommandRequest.",
+						this.getNode().getNodeId()));
+				zwaveCommandClass.handleApplicationCommandRequest(
+						serialMessage, offset+1, 0); //TODO get endpoint from somewhere
```

Comment:
I would probably remove the TODO. As per my email, I don't think it makes any sense to encapsulate a CRC 16 class inside the multi_instance class, since the whole point of CRC16 is to improve reliability, and it doesn't do this if it's not the outer class.